src/v/cloud_topics/objects/level_one_object.h

```diff
@@ -0,0 +1,392 @@
+// containing a series of record batches, followed by a footer with index data.
+//
+// Structure:
+// [Partition 1 Marker][Partition 1 Data][Partition 2 Marker][Partition 2 Data]...[Footer][Footer Size]
+//
+// Components:
+// 1. Partition Marker: A data type delimiter (1 byte) + size (4 bytes) + serialized model::ntp
+//    - Delimiter: 0x01 (data_type::partition_marker)
+//    - Size: uint32_t size of the serialized ntp data
+//    - Data: Serialized model::ntp identifying the partition
+//
+// 2. Partition Data: Sequence of kafka record batches for the partition,
+//    with offsets strictly increasing within each partition
+//    - Each batch prefixed with delimiter 0x00 (data_type::kafka_batch)
+//    - Followed by fixed-width batch header (all header fields in order)
+//    - Followed by batch data (variable size recorded in the header)
+//
+// 3. Footer: A data type delimiter (1 byte) + size (4 bytes) + serialized object_index
+//    - Delimiter: 0x02 (data_type::footer)
+//    - Size: uint32_t size of the serialized object_index data
+//    - Data: Serialized object_index with all partition metadata
+//
+// 4. Footer Size: Final 4 bytes containing uint32_t size of the footer data
+//    (including the delimiter byte and size field)
+//
+// Data type delimiters:
+// - 0x00: kafka_batch - Standard kafka record batch
+// - 0x01: partition_marker - Marks start of new partition
+// - 0x02: footer - Contains object index metadata
+//
+// Index entries are created approximately every 4MiB to enable efficient seeking by
+// offset or timestamp within a partition. The index entries are recorded within the
+// footer.
+//
+// clang-format on
+
+// This file defines the metadata/indexing structure for level one objects. It's
+// written as a footer in the object in cloud storage.
+struct object_index
+  : serde::
+      checksum_envelope<object_index, serde::version<0>, serde::version<0>> {
+    // Information about each partition in the object.
+    struct partition
+      : serde::envelope<partition, serde::version<0>, serde::version<0>> {
+        // The NTP for this partition.
+        model::ntp ntp;
+        // The offset in the file where this partition's data starts (after the
+        // partition marker).
+        size_t file_position = 0;
+
+        struct index_entry
+          : serde::envelope<index_entry, serde::version<0>, serde::version<0>> {
+            // This file_position is the offset in the file where the data batch
+            // that corresponds to this index entry is located.
+            size_t file_position = 0;
+            // The kafka base_offset of the kafka batch that is located at
+            // `file_position`.
+            kafka::offset kafka_offset;
+            // The maximum timestamp seen so far (inclusive of the batch that
+            // generated this index entry) of the kafka batches within the
+            // partition.
+            model::timestamp max_timestamp;
```

Comment:
Great that we're going to enforce monotonicity here with this. Do you know if the semantics here differ from Kafka's when user input timestamps are out of order? I was under the impression the contract was "find the first timestamp >= than `target`, _regardless_ of whether the timestamps were produced in order, but I'm not sure that's correct

Reply:
The indexing isn't bad, the semantics are pretty well defined if not just weird in the face of wildly out of order timestamps.

You need to get the first offset where the timestamp is not before. So keeping the max allows you to prune the space with binary search. If you have an early offset that is way in the future, you just always get that offset back from timequeries.